_studio/mfx_lib/decode/vp9/src/mfx_vp9_dec_decode_hw.cpp

```diff
@@ -85,6 +85,78 @@ bool CheckHardwareSupport(VideoCORE *p_core, mfxVideoParam *p_video_param)
     return true;
 }
 
+class ReferenceFrameStorage
+{
+private:
+    std::vector<UMC_VP9_DECODER::VP9DecoderFrame> m_submittedFrames;
+    mfx_UMC_FrameAllocator *m_frameAllocator;
+
+    void FreeReferenceFrames(UMC::FrameMemID frameId, bool force)
+    {
+        auto find_it = std::find_if(m_submittedFrames.begin(), m_submittedFrames.end(), [frameId](UMC_VP9_DECODER::VP9DecoderFrame frame){ return (frame.currFrame == frameId); });
+        if (find_it != m_submittedFrames.end())
+        {
+            // if force=TRUE need to unlock all reference frames
+            // otherwiseref useless references wil be unlocked in VideoDECODEVP9_HW::UpdateRefFrames()
+            if (force)
```

Comment:
`force` looks like a dirty hack. Consider using of `FreeReferenceFrames` from `UpdateRefFrames`